### DISCUSSION TITLE
Composition testing

### DISCUSSION THREAD

**Discussion by martijn788** [vuejs/test-utils#984] [Q&A]
When you want to test seperate composition-files, without a vue-context, you get warnings about using certain functions outside of the setup function. 

Example: 

`
export function doingMagic(input: string): string {
  // some magic code that uses provide or inject on the background.
} 
`

Example test:
`
it('does some real magic') {
  const data = doingMagic('magical input)
  expect(data).to.equal('output Magic')
}
`

Example console: 
`[Vue warn]: provide() can only be used inside setup().`

So the test passed, but I want to get rid of the error. 
What is the best way to solve this problem?

**Comment by lmiller1990** [comment]
Right now you cannot test something using provide/inject outside of `setup`, since those rely on a Vue instance to inject the value.

If you share your full code, I could make some suggestions.
